ocaml/typing/typecore.ml

```diff
@@ -588,25 +627,27 @@ let escape ~loc ~env ~reason m =
   submode ~loc ~env ~reason m mode_legacy
 
 type expected_pat_mode =
-  { mode : Value.t;
-    tuple_modes : Value.t list; }
+  { mode : Value.l;
+    tuple_modes : Value.l list; }
 
 let simple_pat_mode mode =
-  { mode; tuple_modes = [] }
+  { mode = Value.disallow_right mode; tuple_modes = [] }
 
 let tuple_pat_mode mode tuple_modes =
+  let mode = Value.disallow_right mode in
+  let tuple_modes = List.map Value.disallow_right tuple_modes in
   { mode; tuple_modes }
 
-let allocations : Alloc.t list ref = Local_store.s_ref []
+let allocations : Alloc.r list ref = Local_store.s_ref []
 
 let reset_allocations () = allocations := []
 
 let register_allocation_mode alloc_mode =
-  if not (Alloc.is_const alloc_mode) then
-   allocations := alloc_mode :: !allocations
+  let alloc_mode = Alloc.disallow_left alloc_mode in
```

Comment:
The `if` here is dropped. Is there a reason why?

Reply:
This is to improve performance. The old logic is that, the some allocation is already constant, then we don't need to call `zap_to_ceil` on it later. However, checking for constant involves `zap_to_floor` which is expensive.

The new logic is that, we will call `zap_to_ceil` on all allocations.